### DISCUSSION TITLE
Vox has the prospects to become the best compiler backend

### DISCUSSION THREAD

**Discussion by rempas** [MrSmith33/vox#41] [General]
Of course before we even start, I just want to tell you that I'm closely wanting the progress Vox is making and I'm very happy that you keep doing the extremely hard task. As always, I hope you are enjoying it as much as possible. Now for the actual topic:

From the way I see it, even for people that will not want to use Vox as a straight up compiler and language, it can be the best compiler backend to be used for code generation (of course excluding compiler specific ones)! Let me explain me view behind this idea.

You see, a lot of (non-toy) compilers these days will either use an IR like LLVM, GCC's IR or either compile to another language (mostly C) as their backend. You will rarely see a compiler that outputs either clean assembly or even better, directly create the final executable file (like Vox or TCC). Of course these methods mean that your job to create the backend will be much much easier that if you had to go with the alternative route (which is why I think you are a genius for what you do!) and on top of that, these backends will also do most (if not all) of the optimizations for you so not only you'll get a backend that has an easy "API" but your code will also run very fast. That's awesome right?

Well, this comes with a great cost. Compilation times, the ancient problem! The compilation time of your compiler will always be `x + n`. Where 'n' is the time it takes your compiler to parse and output and 'x' is the time it takes for the "final" compiler to give use the final output. Of course, that's a problem! Most compiler are slow (or bearable at best). So having to parse another language on top of that will make things terrible! Every language that outputs to C is the proof of that. Languages like V, Zig, Nim just to name a few.

Well, at this point, you will want to stop me and tell me that `TCC` exists, right? A fully non-toy C compiler that implements C11 and a couple of GCC extensions. So we can use that right? Yeah, not exactly.... You see, TCC is truly fast. Like.... SUPER fast! So fast that is even 3-4 (or even more) times faster that even if you wrote clean assembly and used an assembler and linker directly. However, there is a problem and that is that TCC doesn't make any optimizations. This includes low level optimizations like the right usage of registers or loop unrolling (and probably a lot more but I'm a n00b in this topic so I don't know) that the high level language cannot make. Or If we could make, It means that we would have to manually write inline assembly and that will be a terrible experience and defeats the whole purpose of using a high level language to make our lives easier and in that case, it would be better to just output assembly directly.

But wait, Vox exists! Vox is compiler that is close to TCC when it comes to how fast it complies code! I tried to compare GCC, LDC and TCC with Vox to create a program that just prints the string: "hello, word!" and exist (using inline assembly for the other 3 compilers) and the results were:

```
GCC: 0.033s
LDC: 0.066s
TCC: 0.002s
VOX: 0.003s
```

Of course, I know that this is the smallest possible program and the results would be different in a bigger real life program but still, this is enough to prove my point!

And the difference with TCC is that you want to actually make optimizations (from what you have told me) and I suppose that you will make the important ones that I mentioned so there should be no problem! Vox can become the best compile backend out there! So good that it will be worth for the current languages to change their current backend (especially if it is the ugly LLVM) to Vox despite the fact that this will be a time consuming and hard task to do!

Now, I'm not just saying this because I'm bored and I want to start a conversation. I'm actually saying it because I really think that you should actually start promoting that! You have explained to me that Vox was created mainly for Game dev and if you don't plan to make Vox into a language with modern features that languages that can compete against language like Rust, Python, Java, and D have then I think that you should also start advertising the power than the Vox compiler has itself and how it can be used for compiler developers so people can create whatever they want. I know because I'm actually trying to make one and not having to learn assembly and having the power of a language as a backend can seven me tons of time on actually focusing to the language itself rather than on how to do some complicated stuff like learning how to properly parse expressions and re-order them (just to name what I'm currently trying to make).

I also think having a shared backend makes more sense for compiler developers as this means that people that want to improve it will contribute to the project so everyone wins for every improvement that the backend will get (the same idea with LLVM and why so many people choose it). So it will also help the compiler attract more contributors and get better!

What do you think about all that? I'm I just dreaming too much or are my points valid?

**Comment by MrSmith33** [comment]
You are correct on all points and this is already in my plans. The main difficulty with that is that I need to separate backend so it is standalone and has C API/ABI available. This is non-trivial and requires some serious motivation. Multiple people expressed the desire to use such backend if it was available, but I need someone who is more invested in trying out Vox backend to give me proper motivation.

**Reply by rempas** [reply]
Separating the backend will require a lot of extra work. If I understand what you are talking about, this means creating an IR (similar to GCC and Clang) so there is no need to have Vox as a whole compiler and won't have to generate vox files and then re-parse them again (so we avoid 1 step). While this will allow us to have a little bit (or maybe not only a little bit?) faster compile times and it will allow us to have the backend as a library, I don't see a lot of advantages. Unless you mean something else and I didn't understand or if there are other things that I haven't thought about.

However, sure! If you think that this makes more sense and it will help in structuring the project better. However, I think that even at that point, Vox is ready to go! You just generate Vox soure (similar to how most compilers generate C source) and you let the compiler do its job. The C ABI and linkage will be a problem and we will also need to be able to both link and produce shared object files (.so) on Unix but these things can wait for now. After these two have been implemented, you can start designing a new structure and separate the backend unless have you changed your mind until then and you decide that it is not worth it. As I thought about making a compiler that will have a seperate bakcend too, I thought about a great way of doing it! Of course, I can share it with you when the time comes if you want to hear it.

Also, keep in mind that if you create an IR API, it must be as high level as possible so it doesn't defeat the purpose of doing the compiler designer's life easier. For example, I consider [QBE](https://c9x.me/compile/)'s [IR](https://c9x.me/compile/doc/il.html) to be close to assembly so if someone told me to use this, I would say that it doesn't worth the trouble and I would go to directly to assembly.

So while I think that having a separate backend and create an IR API would be the next-level design, I'm also realistically saying that like you said, it will need a lot of work and motivation. So I think that it is wise to think if it's really worth it and we should probably focus on other things that needs to be implemented.

With that been said, I know that I have posted so many stuff here and I always talked about theory and never helped.... BUT! I'm glad to say that I'm now ready to help! You see, like I said, I'm already working in a frontend. It's my first time and it's not something very big right now but I'll be glad to use Vox as the backend and I would like to give you feedback and answer to any question you have. I may even re-write it in Vox if it is possible (and btw, please update the spec as soon as you can). I also got more motivated on working in a frontend now that I thought about using Vox as a backend. Ok, that's because this will make my life easier and I won't have to work so hard but but but... I'm beating laziness little be little so compared to nothing, that's still better right?

When it comes to motivation, keep in mind that I tried learning ELF files and machine instructions but I found it very hard to even make it work (and trust me, I really tried). I'm pretty sure that a lot of other have tried and had the same results. However, you are able to create a fully industry level (CROSS-PLATFORM!!!!) compiler!!! No-one else has made something like Vox before from what I know (expect for some small C compilers like TCC but I explained you the problems with it). So while I cannot guarantee and predict the future, It's almost sure that Vox will EXPLODE if people learn it and start using it. Both as a whole language and as a backend to use for a frontend!

To add on that, given the fact how much you care and how hard you work (and of course your very rare knowledge), I think that the any limitations that will get crossed will be solved. Like you said, a lot of people have expressed the need of finding and using a backend like Vox. And again, Vox is ready to be used NOW! Even some months ago, I was SHOCKED (you remember my posts right?) about how powerful Vox is. I didn't believed it how people haven't found that gem yet! And Vox has come a long way since them. And even now, our talks and ideas are a progress!

If motivation is your problem, I think I covered you. Please feel free to discuss any idea with me and If I can help with any way, I will be glad to do so! :)

**Reply by MrSmith33** [reply]
ok, hop on to a [discord server](https://discord.gg/HpYYhH4) where I have a channel for Vox

**Reply by rempas** [reply]
Nice! I'm in! Do you need anything specific for now?

**Reply by MrSmith33** [reply]
Are you writing your frontend in C?

**Reply by rempas** [reply]
I'm writing it in D using "-betterC".

**Reply by MrSmith33** [reply]
The way I see Vox being used from languages is by linking either with static lib or with dynamic lib. In order to actually use it, that language should define bindings. I think I can try doing this with `-betterC` example D program, that wouldn't import vox files, but will use extern(C) API + lib.

**Reply by rempas** [reply]
Good morning! Are you talking about the shared object thing I mentioned? In that case, yeah! Just like D does it with "extern(C)" but without the "ImportC" which means that we will have to manually create bindings. That's not something tragic however, it's how most languages do it anyway. In in the future, we can create a program that will do this for us by parsing C header files. But for now, just making the shared object linking work on ELF files will be enough imo. I also tried to convert the code for my frontend from D to Vox yesterday but I remembered that I cannot link with libc as I was doing that. So here is a practical example ;)

Also, please! Give the ability to declare/call a symbol under a custom name. This will allow us to use libraries for other languages that have their own mangle without requiring Vox to know the mangle of these languages. Just have that in mind, it's not an urgently.

Static libraries (.a) will be awesome but if you ask me, first focus on making shared objects work and then you can make static libraries work. I'm saying that because shared objects are more common from what I know (and this is also true for dynamic libraries on Windows too). And some libraries like "libc" (at least on GNU/Linux distros) only offer a shared version.

P.S. Thank you for updating the spec!!! :heart:
